File: heart/src/main/java/com/vantagecircle/heartrate/fragment/ui/HeartFragment.java
package com.vantagecircle.heartrate.fragment.ui;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.vantagecircle.heartrate.HeartApplication;
import com.vantagecircle.heartrate.R;
import com.vantagecircle.heartrate.component.DaggerFragmentComponent;
import com.vantagecircle.heartrate.component.FragmentComponent;
import com.vantagecircle.heartrate.core.PulseSupport;
import com.vantagecircle.heartrate.data.DataManager;
import com.vantagecircle.heartrate.databinding.HeartRateLayoutBinding;
import com.vantagecircle.heartrate.fragment.BaseFragment;
import com.vantagecircle.heartrate.fragment.presenter.HeartFragmentPresenter;
import com.vantagecircle.heartrate.model.Heart;
import com.vantagecircle.heartrate.module.FragmentModule;

import javax.inject.Inject;

public class HeartFragment extends BaseFragment {
    private static final String TAG = HeartFragment.class.getSimpleName();
    @Inject
    PulseSupport mPulseSupport;
    @Inject
    DataManager mDataManager;
    @Inject
    Heart mHeart;

    private FragmentComponent mFragmentComponent;
    private HeartRateLayoutBinding mHeartRateLayoutBinding;
    private HeartFragmentPresenter mHeartFragmentPresenter;

    public static HeartFragment newInstance() {
        return new HeartFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        Log.e(TAG, "onCreate");
        super.onCreate(savedInstanceState);
    }

    @Override
    protected void setFragmentComponent() {
        if (mFragmentComponent == null) {
            mFragmentComponent = DaggerFragmentComponent.builder()
                    .fragmentModule(new FragmentModule(this))
                    .heartComponent(HeartApplication.get(this.getContext()).getHeartComponent())
                    .build();
        }
        mFragmentComponent.inject(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        Log.e(TAG, "onCreateView");
        mHeartRateLayoutBinding = DataBindingUtil.inflate(inflater, R.layout.heart_rate_layout, container, false);
        return mHeartRateLayoutBinding.getRoot();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        Log.e(TAG, "onActivityCreated");
        super.onActivityCreated(savedInstanceState);
        init();
    }

    @Override
    protected void init() {
        mHeartFragmentPresenter = new HeartFragmentPresenter(this.getActivity(), mPulseSupport, mDataManager);
        mHeartFragmentPresenter.initialize(mHeart);
        mHeartRateLayoutBinding.setHeartPresenter(mHeartFragmentPresenter);
    }

    @Override
    public void onPause() {
        Log.e(TAG, "onPause");
        super.onPause();
        if (mHeartFragmentPresenter != null)
            mHeartFragmentPresenter.stop();
    }
}

File: app/src/main/java/com/bapidas/heartrate/ui/fragment/ui/HeartFragment.kt
package com.bapidas.heartrate.ui.fragment.ui

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import com.bapidas.heartrate.HeartApplication.Companion.get
import com.bapidas.heartrate.R
import com.bapidas.heartrate.core.HeartRate
import com.bapidas.heartrate.data.DataManager
import com.bapidas.heartrate.databinding.HeartRateLayoutBinding
import com.bapidas.heartrate.di.component.DaggerFragmentComponent
import com.bapidas.heartrate.di.component.FragmentComponent
import com.bapidas.heartrate.di.module.FragmentModule
import com.bapidas.heartrate.ui.base.BaseFragment
import com.bapidas.heartrate.ui.fragment.presenter.HeartFragmentPresenter
import javax.inject.Inject

class HeartFragment : BaseFragment() {
    @Inject
    lateinit var heartRate: HeartRate

    @Inject
    lateinit var mDataManager: DataManager

    private var mFragmentComponent: FragmentComponent? = null
    private var mHeartRateLayoutBinding: HeartRateLayoutBinding? =
        null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        mHeartRateLayoutBinding = DataBindingUtil.inflate(
            inflater, R.layout.heart_rate_layout,
            container, false
        )
        return mHeartRateLayoutBinding?.root
    }

    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?
    ) {
        super.onViewCreated(view, savedInstanceState)
        init()
    }

    override fun setFragmentComponent(context: Context) {
        if (mFragmentComponent == null) {
            mFragmentComponent = DaggerFragmentComponent.builder()
                .fragmentModule(FragmentModule(this))
                .appComponent(get(context).appComponent)
                .build()
        }
        mFragmentComponent?.inject(this)
    }

    override fun init() {
        val mHeartFragmentPresenter = HeartFragmentPresenter(
            activity,
            heartRate.heartSupport, mDataManager
        )
        mHeartRateLayoutBinding?.heartPresenter = mHeartFragmentPresenter
    }

    companion object {
        private val TAG = HeartFragment::class.java.simpleName
        fun newInstance(): HeartFragment {
            return HeartFragment()
        }
    }
}